### ISSUE TITLE
Vulkano shouldn't link to MoltenVK by default on macOS

### ISSUE THREAD

**Issue by nlordell** [vulkano-rs/vulkano#928]
OK, so this is going to be a long block of text, and some code (there is a TL;DR at the end). With the open source release of MoltenVK and the 1.0.69 release of the macOS LunarG Vulkan SDK I believe that the linking by default to the MoltenVK framework should no longer happen.

On Windows, Linux and Android, a shared object is loaded and the `vkGetInstanceProcAddr` symbol is loaded from it. They rely on `vulkan.{dll,so}` being available to load. However, through the `Loader` trait it allows crate users to load `vkGetInstanceProcAddr` however they see fit. That means that if I want to add `println!("cargo:rustc-link-lib=vulkan");` to my `build.rs` and provide `vkGetInstanceProcAddr` I can (I think... didn't try it to be honest).

As of the  LunarG Vulkan SDK 1.0.69 for macOS there is a `libvulkan.1.dylib` loader that acts more like its Windows, Linux and Android counterparts (more on that later). That means that if `libvulkan.1.dylib` is in an expected place, then the `DynamicLibraryLoader` implementation of `Loader` will work. If we choose to go down that path we are still linking to MoltenVK (and other frameworks) even if it is no longer needed.

But why not just use the MoltenVK framework? The LunarG Vulkan SDK loader has a bunch of cool features that you might want (eg. support for `VK_LAYER_PATH` and `VK_INSTANCE_LAYERS` environment variables). You can take a look at this gist: https://gist.github.com/nlordell/184d363cffe0d7895155ee5fee89aa9a
The important bit is the output of running with dynamic loading of `libvulkan.1.dylib` vs linking against `MoltenVK`:
```
$ export VK_LAYER_PATH="/path/to/layer/configs"
$ cargo run # with dynamic loading of libvulkan.1.dylib
   Compiling mvk v0.1.0 (file:///Users/nlordell/Developer/rust/mvk)
    Finished dev [unoptimized + debuginfo] target(s) in 1.51 secs
     Running `target/debug/mvk`
VK_LAYER_GOOGLE_unique_objects
VK_LAYER_GOOGLE_threading
VK_LAYER_LUNARG_standard_validation
VK_LAYER_LUNARG_core_validation
VK_LAYER_LUNARG_parameter_validation
VK_LAYER_LUNARG_object_tracker
$ cargo run --features "framework-linking" # with linking against MoltenVK
   Compiling mvk v0.1.0 (file:///Users/nlordell/Developer/rust/mvk)
    Finished dev [unoptimized + debuginfo] target(s) in 0.90 secs
     Running `target/debug/mvk`
MoltenVK
```

In terms of code breakage, using the dynamic loading of `libvulkan.1.dylib` can be gated behind a feature to prevent breakage. Additionally, if the method of linking to MoltenVK gets completely removed, crate users that prefer to use MoltenVK framework can always add the `cargo:rustc-link-lib=...` lines to their `build.rs` and implement a `StaticallyLinkedVulkanLoader` implementation to the `Loader` trait.

TL;DR: LunarG Vulkan SDK provides a vulkan loader analogous to its Windows, Linux and Android counterpart that is "better" than MoltenVK framework - Vulkano shouldn't force crate users to link against a framework that is no longer needed and provide the ability to use the vulkan loader instead. It will also be more consistent with what Vulkano does in Windows, Linux and Android.

PS: I have no idea about iOS, but according to the docs there is currently no vulkan loader library, so maybe it should still link to MoltenVK framework.

**Comment by nlordell** [comment]
The triangle example works with using the `DynamicLibraryLoader` with `"libvulkan.1.dylib"` from the LunarG Vulkan SDK and not linking to the MoltenVK framework:

<img width="1024" alt="image" src="https://user-images.githubusercontent.com/4210206/36865805-e038609e-1d90-11e8-8f1c-f60ef16779f0.png">

**Comment by tomaka** [comment]
I totally agree with that. The reason why it's automatically linked today is purely for convenience.

**Comment by nlordell** [comment]
Would you be willing to accept a PR for defaulting to using `DynamicLibraryLoader` with the appropriate documentation on how to use MoltenVK and `StaticallyLinkedVulkanLoader`? I understand if the answer is "no" as this could pentially break how a lot of people currently build.

The other option is to create a `no-moltenvk-linking` feature that if enabled will use `DynamicLibraryLoader` and `libvulkan.1.dylib`. Either way, I don't mind submitting a PR, just let me know how you prefer it. Just want to be able to use `libvulkan.1.dylib` somehow :D.

**Comment by JustAPerson** [comment]
@tomaka I'm curious if you've thought more about this. I'd also be happy to try helping.
Using the LunarG SDK would be very helpful.

Additionally, @nlordell's gist shows the proper `build.rs` to work for static linking, which I didn't find documented anywhere obvious. Would you like a PR to document that somewhere?

**Comment by gibfahn** [comment]
I think this was fixed by https://github.com/vulkano-rs/vulkano/pull/948

**Comment by rukai** [comment]
Yep, looks like it
